src/OrbitBase/include/OrbitBase/StopToken.h

```diff
@@ -36,6 +40,34 @@ class StopToken {
   orbit_base::Future<void> future_;
 };
 
+// Takes a `Future<T>` and a `StopToken` and returns a `Future<CanceledOr<T>>`. The returned future
+// will complete when either the input future completes or when the StopToken indicates that a stop
+// was requested - whatever happens first. When a stop was requested the resulting value will be
+// marked as `Canceled`, otherwise it will contain the `T`.
+//
+// Note that this doesn't mean that the operation that was supposed to produce the result of
+// `Future<T>` will be interrupted. It will finish as usual, but the result will not be used
+// by this code path.
+// To make a background job truly cancelable you have to make the task take a StopToken instead.
+//
+// This function is still useful though to make an uninterruptable task "cancelable".
+template <typename T>
+auto WhenValueOrCanceled(const Future<T>& value, const StopToken& stop_token)
+    -> Future<CanceledOr<T>> {
+  orbit_base::ImmediateExecutor executor{};
+  return orbit_base::WhenAny(value, stop_token.GetFuture())
+      .Then(&executor,
+            [](const std::variant<orbit_base::VoidToMonostate_t<T>, std::monostate>& result)
+                -> CanceledOr<T> {
+              if (result.index() == 1) return outcome::failure(Canceled{});
```

Comment:
I think this deserves a comment. What is index 1?

Reply:
Added a comment. Thx.